ember-server/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala

```diff
@@ -151,18 +153,26 @@ private[server] object ServerHelpers {
         }
         .drain
 
-    sg.server[F](bindAddress, additionalSocketOptions = additionalSocketOptions)
-      .interruptWhen(shutdown.signal.attempt)
-      // Divorce the scopes of the server stream and handler streams so the
-      // former can be terminated while handlers complete.
-      .prefetch
-      .map { connect =>
-        shutdown.trackConnection >>
-          Stream
-            .resource(connect.flatMap(upgradeSocket(_, tlsInfoOpt)))
-            .flatMap(withUpgradedSocket(_))
+    Stream
+      .resource(
+        sg.serverResource[F](bindAddress, additionalSocketOptions = additionalSocketOptions))
+      .attempt
```

Comment:
I added a combinator called `onComplete` to Cats, which is basically attempt + evalTap + rethrow but I don't think it's come downstream yet. If it has, FS2 also has an `onComplete` so it's probably getting shadowed